IllegalStateExceptions about no UnitOfWork on the thread

On z/OS the customer is reporting `IllegalStateException: <cut> is not associated with the unit of work trying to suspend it:`

Looking at the logs, we see that this was never suspended on the relevant thread, yet it attempts to resume. 

Customer environment
```
product = WAS FOR Z/OS 25.0.0.11 (wlp-1.0.107.cl251120251020-0302)
java.version = 21.0.8
java.runtime = IBM Semeru Runtime Certified Edition for z/OS (21.0.8+9)
os = z/OS (03.01.00; s390x) (en_US)
```

Error stack shows
```
[3/6/26, 13:03:58:040 UTC] 00012208 com.ibm.ejs.container.BusinessExceptionMappingStrategy E CNTR0019E: EJB threw an unexpected (non-declared) exception during invocation of method "call". Exception data: com.ibm.websphere.csi.CSIException: Tran Strategy preInvoke failed; nested exception is:
java.lang.IllegalStateException: Context [ContextImpl [Token: e25672067d09bd4d060000000a967e28, RegistryToken: c2c2c7e9d9c5c7e30000000100000040000000007e1730400000002a000000000200000000000000000000000000000000000000000000000000000000000000, State: ACTIVE, UOWCoordinator: com.ibm.tx.jta.embeddable.impl.EmbeddableTransactionImpl@185bfa34#tid=155143092,active=1,suspended=0thread=00012208,gtrid=0000019cc33f65a800000001093f4bb47ce176de29431cc155bb8f50fa344c177d1509f8]] is not associated with the unit of work trying to suspend it: com.ibm.tx.ltc.embeddable.impl.EmbeddableLocalTranCoordImpl@c8d877c0;SUSPENDED;
at com.ibm.ejs.csi.TransactionControlImpl.preInvoke(TransactionControlImpl.java:259)
at com.ibm.ejs.container.EJSContainer.preInvokeActivate(EJSContainer.java:2943)
at com.ibm.ejs.container.EJSContainer.EjbPreInvoke(EJSContainer.java:2392)
at com.hlag.fis.legacy.core.domain.gateway.EJSRemote0SLGenericGenActionBlockGateway_5da3b206.call(EJSRemote0SLGenericGenActionBlockGateway_5da3b206.java)
at com.hlag.fis.legacy.core.domain.gateway._EJSRemote0SLGenericGenActionBlockGateway_5da3b206_Tie.call(_EJSRemote0SLGenericGenActionBlockGateway_5da3b206_Tie.java:1)
at com.hlag.fis.legacy.core.domain.gateway._EJSRemote0SLGenericGenActionBlockGateway_5da3b206_Tie._invoke(_EJSRemote0SLGenericGenActionBlockGateway_5da3b206_Tie.java)
at org.apache.yoko.orb.OBPortableServer.ServantDispatcher.dispatch(ServantDispatcher.java:189)
at org.apache.yoko.orb.OBPortableServer.POA_impl._OB_dispatch(POA_impl.java:1245)
at org.apache.yoko.orb.OB.DispatchRequest_impl.invoke(DispatchRequest_impl.java:57)
at com.ibm.ws.transport.iiop.yoko.ExecutorDispatchStrategy$1.run(ExecutorDispatchStrategy.java:44)
at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:344)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
at java.base/java.lang.Thread.run(Thread.java:1595)
Caused by: java.lang.IllegalStateException: Context [ContextImpl [Token: e25672067d09bd4d060000000a967e28, RegistryToken: c2c2c7e9d9c5c7e30000000100000040000000007e1730400000002a000000000200000000000000000000000000000000000000000000000000000000000000, State: ACTIVE, UOWCoordinator: com.ibm.tx.jta.embeddable.impl.EmbeddableTransactionImpl@185bfa34#tid=155143092,active=1,suspended=0thread=00012208,gtrid=0000019cc33f65a800000001093f4bb47ce176de29431cc155bb8f50fa344c177d1509f8]] is not associated with the unit of work trying to suspend it: com.ibm.tx.ltc.embeddable.impl.EmbeddableLocalTranCoordImpl@c8d877c0;SUSPENDED;
at com.ibm.ws.zos.tx.internal.ContextManagerImpl.suspend(ContextManagerImpl.java:390)
at com.ibm.ws.zos.tx.internal.NativeTransactionManager.UOWEvent(NativeTransactionManager.java:1791)
at com.ibm.tx.ltc.impl.LocalTranCurrentImpl.invokeEventListener(LocalTranCurrentImpl.java:481)
at com.ibm.tx.ltc.impl.LocalTranCurrentImpl.suspend(LocalTranCurrentImpl.java:184)
at com.ibm.tx.ltc.impl.LocalTranCurrentSet.suspend(LocalTranCurrentSet.java:159)
at com.ibm.ejs.csi.TranStrategy.suspendLocalTx(TranStrategy.java:493)
at com.ibm.ejs.csi.Required.preInvoke(Required.java:52)
at com.ibm.ejs.csi.TransactionControlImpl.preInvoke(TransactionControlImpl.java:223)
... 13 more
```

--- a/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
+++ b/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
@@ -87,6 +87,12 @@ public class TransactionContextImpl implements ThreadContext {
     public ThreadContext clone() {
         try {
             TransactionContextImpl copy = (TransactionContextImpl) super.clone();
+            // Copy the ThreadLocal value to the cloned instance
+            copy.suspendedUOW = new ThreadLocal<UOWToken>();
+            UOWToken currentValue = this.suspendedUOW.get();
+            if (currentValue != null) {
+                copy.suspendedUOW.set(currentValue);
+            }
             return copy;
         } catch (CloneNotSupportedException x) {
             throw new RuntimeException(x);

--- a/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
+++ b/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -67,7 +67,7 @@ public class TransactionContextImpl implements ThreadContext {
     /**
      * Unit of work that was on the thread of execution prior to invoking the contextual task.
      */
-    private transient UOWToken suspendedUOW;
+    private transient ThreadLocal<UOWToken> suspendedUOW = new ThreadLocal<UOWToken>();
 
     /**
      * Indicates that prior to invoking a task, any transaction that is present on the thread of execution
@@ -104,7 +104,8 @@ public class TransactionContextImpl implements ThreadContext {
             // Suspend whatever is currently on the thread.
             try {
                 UOWManager uowManager = UOWManagerFactory.getUOWManager();
-                suspendedUOW = uowManager.suspend();
+                UOWToken localuSpendedUOW = uowManager.suspend();
+                suspendedUOW.set(localuSpendedUOW);
             } catch (com.ibm.ws.uow.embeddable.SystemException e) {
             }
 
@@ -156,10 +157,11 @@ public class TransactionContextImpl implements ThreadContext {
 
             // Resume the original transaction.
             try {
-                if (suspendedUOW != null) {
+                UOWToken localSuspendedUOW = suspendedUOW.get();
+                if (localSuspendedUOW != null) {
                     UOWManager uowManager = UOWManagerFactory.getUOWManager();
-                    uowManager.resume(suspendedUOW);
-                    suspendedUOW = null;
+                    uowManager.resume(localSuspendedUOW);
+                    suspendedUOW.remove();
                 }
             } catch (Throwable e) {
                 exception = e;

--- a/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
+++ b/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
@@ -187,6 +187,8 @@ public class TransactionContextImpl implements ThreadContext {
      * @throws ClassNotFoundException
      */
     private void readObject(java.io.ObjectInputStream in) throws IOException, ClassNotFoundException {
+        suspendedUOW = new ThreadLocal<UOWToken>();//Reinitialize transient field.
+
         GetField fields = in.readFields();
 
         Object type = fields.get(TYPE, null);

--- a/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
+++ b/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2022 IBM Corporation and others.
+ * Copyright (c) 2012, 2026 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -7,8 +7,6 @@
  *
  * SPDX-License-Identifier: EPL-2.0
  *
- * Contributors:
- *     IBM Corporation - initial API and implementation
  *******************************************************************************/
 package com.ibm.ws.transaction.context.internal;
 

